// @serenity-js/webdriverio/esm/adapter/reporter/TagPrinter.js
import { BrowserTag, PlatformTag } from '@serenity-js/core/lib/model/index.js';
/**
 * @package
 */
export class TagPrinter {
    tagsFor(capability) {
        const desiredCapabilities = this.desired(capability);
        return [
            this.browserTagFor(desiredCapabilities),
            this.platformTagFor(desiredCapabilities),
        ];
    }
    browserTagFor(capabilities) {
        return new BrowserTag(this.browserNameFrom(capabilities), this.browserVersionFrom(capabilities));
    }
    platformTagFor(capabilities) {
        return new PlatformTag(this.platformNameFrom(capabilities), this.platformVersionFrom(capabilities));
    }
    browserNameFrom(capabilities) {
        return capabilities.browserName
            || capabilities.browser
            || (capabilities['appium:app'] && capabilities['appium:app'].replace('sauce-storage:', ''))
            || 'unknown';
    }
    browserVersionFrom(capabilities) {
        return capabilities.deviceName // mobile web
            || capabilities['appium:deviceName']
            || capabilities.browserVersion // W3C format
            || capabilities.version // JSONWP format
            || capabilities.browser_version; // BrowserStack
    }
    platformNameFrom(capabilities) {
        return capabilities.platformName
            || capabilities['appium:platformName']
            || capabilities.platform
            || capabilities.os
            || 'unknown';
    }
    platformVersionFrom(capabilities) {
        return capabilities['appium:platformVersion']
            || capabilities.os_version;
    }
    desired(capabilities) {
        return this.isW3C(capabilities)
            ? capabilities.alwaysMatch
            : capabilities;
    }
    isW3C(capabilities) {
        return !!capabilities.alwaysMatch;
    }
}
