// @serenity-js/webdriverio/esm/screenplay/models/WebdriverProtocolErrorCode.js
/**
 * https://developer.mozilla.org/en-US/docs/Web/WebDriver/Errors#table_of_errors
 * https://github.com/SeleniumHQ/selenium/blob/e70e22390000a65a6b706fbb33bbff8913555670/javascript/node/selenium-webdriver/lib/error.js#L441
 *
 * @package
 */
export var WebdriverProtocolErrorCode;
(function (WebdriverProtocolErrorCode) {
    WebdriverProtocolErrorCode["WebDriverError"] = "unknown error";
    WebdriverProtocolErrorCode["DetachedShadowRootError"] = "detached shadow root";
    WebdriverProtocolErrorCode["ElementClickInterceptedError"] = "element click intercepted";
    WebdriverProtocolErrorCode["ElementNotInteractableError"] = "element not interactable";
    WebdriverProtocolErrorCode["ElementNotSelectableError"] = "element not selectable";
    WebdriverProtocolErrorCode["InsecureCertificateError"] = "insecure certificate";
    WebdriverProtocolErrorCode["InvalidArgumentError"] = "invalid argument";
    WebdriverProtocolErrorCode["InvalidCookieDomainError"] = "invalid cookie domain";
    WebdriverProtocolErrorCode["InvalidCoordinatesError"] = "invalid coordinates";
    WebdriverProtocolErrorCode["InvalidElementStateError"] = "invalid element state";
    WebdriverProtocolErrorCode["InvalidSelectorError"] = "invalid selector";
    WebdriverProtocolErrorCode["NoSuchSessionError"] = "invalid session id";
    WebdriverProtocolErrorCode["JavascriptError"] = "javascript error";
    WebdriverProtocolErrorCode["MoveTargetOutOfBoundsError"] = "move target out of bounds";
    WebdriverProtocolErrorCode["NoSuchAlertError"] = "no such alert";
    WebdriverProtocolErrorCode["NoSuchCookieError"] = "no such cookie";
    WebdriverProtocolErrorCode["NoSuchElementError"] = "no such element";
    WebdriverProtocolErrorCode["NoSuchFrameError"] = "no such frame";
    WebdriverProtocolErrorCode["NoSuchShadowRootError"] = "no such shadow root";
    WebdriverProtocolErrorCode["NoSuchWindowError"] = "no such window";
    WebdriverProtocolErrorCode["ScriptTimeoutError"] = "script timeout";
    WebdriverProtocolErrorCode["SessionNotCreatedError"] = "session not created";
    WebdriverProtocolErrorCode["StaleElementReferenceError"] = "stale element reference";
    WebdriverProtocolErrorCode["TimeoutError"] = "timeout";
    WebdriverProtocolErrorCode["UnableToSetCookieError"] = "unable to set cookie";
    WebdriverProtocolErrorCode["UnableToCaptureScreenError"] = "unable to capture screen";
    WebdriverProtocolErrorCode["UnexpectedAlertOpenError"] = "unexpected alert open";
    WebdriverProtocolErrorCode["UnknownCommandError"] = "unknown command";
    WebdriverProtocolErrorCode["UnknownMethodError"] = "unknown method";
    WebdriverProtocolErrorCode["UnsupportedOperationError"] = "unsupported operation";
})(WebdriverProtocolErrorCode || (WebdriverProtocolErrorCode = {}));
