// @serenity-js/serenity-bdd/lib/stage/crew/serenity-bdd-reporter/processors/transformations/activityStarted.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.activityStarted = activityStarted;
/**
 * @package
 */
function activityStarted(activityId, name, startedAt) {
    return (context) => {
        const step = {
            number: context.steps.size + 1,
            /**
             * It doesn't look like the TestStep.level was used anywhere in the Serenity BDD Reporter,
             * https://github.com/serenity-bdd/serenity-core/blob/5bebe8e77cf0c4ea99e0d3d1035d54822bcde9af/serenity-model/src/main/java/net/thucydides/model/domain/TestStep.java
             */
            // level: 0,
            // precondition: false,
            description: name.value,
            startTime: startedAt.toISOString(),
            children: [],
            reportData: [],
            screenshots: [],
            duration: 0,
            result: undefined,
        };
        context.steps.set(activityId.value, {
            step,
            parentActivityId: context.currentActivityId,
        });
        context.currentActivityId = activityId;
        return context;
    };
}
