// @serenity-js/serenity-bdd/lib/stage/crew/serenity-bdd-reporter/processors/transformations/scenarioDetailsOf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scenarioDetailsOf = scenarioDetailsOf;
const tiny_types_1 = require("tiny-types");
const mappers_1 = require("../mappers");
/**
 * @package
 */
function scenarioDetailsOf(details) {
    return (context) => {
        const name = (0, tiny_types_1.ensure)('scenario name', details.name.value, (0, tiny_types_1.isNotBlank)());
        const category = (0, tiny_types_1.ensure)('scenario category', details.category.value, (0, tiny_types_1.isNotBlank)());
        const storyName = (0, mappers_1.escapeHtml)(category);
        const requirementsHierarchy = context.requirementsHierarchy.hierarchyFor(details.location.path);
        const pathElements = requirementsHierarchy.map(name => ({
            name,
            description: humanReadable(name),
        }));
        context.report.name = (0, mappers_1.escapeHtml)(name);
        context.report.title = (0, mappers_1.escapeHtml)(name);
        context.report.manual = false;
        context.report.testSteps = [];
        context.report.userStory = {
            id: (0, mappers_1.dashify)(category),
            storyName: storyName,
            displayName: storyName,
            path: requirementsHierarchy.join('/'),
            type: 'feature',
            narrative: '',
            pathElements,
        };
        return context;
    };
}
function humanReadable(name) {
    const result = name
        .split('_')
        .join(' ');
    return result.charAt(0).toUpperCase() + result.slice(1);
}
