// @serenity-js/web/lib/screenplay/models/Locator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Locator = void 0;
const core_1 = require("@serenity-js/core");
const io_1 = require("@serenity-js/core/lib/io");
const util = __importStar(require("util")); // eslint-disable-line unicorn/import-style
const RootLocator_1 = require("./RootLocator");
const selectors_1 = require("./selectors");
/**
 * [`Locator`](https://serenity-js.org/api/web/class/Locator/) uses a [`Selector`](https://serenity-js.org/api/web/class/Selector/) to locate a [`PageElement`](https://serenity-js.org/api/web/class/PageElement/) or [`PageElement`](https://serenity-js.org/api/web/class/PageElements/)
 * within the [`Page`](https://serenity-js.org/api/web/class/Page/).
 *
 * ## Learn more
 * - [`RootLocator`](https://serenity-js.org/api/web/class/RootLocator/)
 * - [`Page.locate`](https://serenity-js.org/api/web/class/Page/#locate)
 * - [`PageElement`](https://serenity-js.org/api/web/class/PageElement/)
 * - [`PageElement`](https://serenity-js.org/api/web/class/PageElements/)
 *
 * @group Models
 */
class Locator extends RootLocator_1.RootLocator {
    parent;
    selector;
    constructor(parent, selector) {
        super();
        this.parent = parent;
        this.selector = selector;
    }
    async switchToFrame(element) {
        await this.parent.switchToFrame(element);
    }
    async switchToParentFrame() {
        await this.parent.switchToParentFrame();
    }
    async switchToMainFrame() {
        await this.parent.switchToMainFrame();
    }
    /**
     * Expresses [`ByCss`](https://serenity-js.org/api/web/class/ByCss/), [`ById`](https://serenity-js.org/api/web/class/ById/), or [`ByTagName`](https://serenity-js.org/api/web/class/ByTagName/) as a [`ByCss`](https://serenity-js.org/api/web/class/ByCss/) selector.
     *
     * @throws LogicError
     *  if the `selector` can't be expressed as [`ByCss`](https://serenity-js.org/api/web/class/ByCss/)
     *
     * @param selector
     * @protected
     */
    asCssSelector(selector) {
        if (selector instanceof selectors_1.ByCss) {
            return selector;
        }
        if (selector instanceof selectors_1.ById) {
            return new selectors_1.ByCss(`#${selector.value}`);
        }
        if (selector instanceof selectors_1.ByTagName) {
            return new selectors_1.ByCss(`${selector.value}`);
        }
        throw new core_1.LogicError(`${selector} can't be expressed as a CSS selector`);
    }
    toString() {
        return this.selector.toString();
    }
    /**
     * Custom [Node.js inspection method](https://nodejs.org/api/util.html#utilinspectcustom).
     */
    [util.inspect.custom] = (0, io_1.inspectedObject)(this, ['parent', 'selector']);
}
exports.Locator = Locator;
