// @serenity-js/serenity-bdd/lib/stage/crew/serenity-bdd-reporter/SerenityBDDReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SerenityBDDReporter = void 0;
const core_1 = require("@serenity-js/core");
const events_1 = require("@serenity-js/core/lib/events");
const io_1 = require("@serenity-js/core/lib/io");
const model_1 = require("@serenity-js/core/lib/model");
const tiny_types_1 = require("tiny-types");
const processors_1 = require("./processors");
/**
 * A [`StageCrewMember`](https://serenity-js.org/api/core/interface/StageCrewMember/) that produces [Serenity BDD](http://serenity-bdd.info/)-standard JSON reports
 * to be parsed by [Serenity BDD CLI Reporter](https://github.com/serenity-bdd/serenity-cli)
 * to produce HTML reports and living documentation.
 *
 * ## Registering Serenity BDD Reporter programmatically
 *
 * ```ts
 * import { ArtifactArchiver, configure } from '@serenity-js/core';
 * import { SerenityBDDReporter } from '@serenity-js/serenity-bdd';
 *
 * configure({
 *   crew: [
 *     ArtifactArchiver.fromJSON({
 *       outputDirectory: './target/site/serenity'
 *     }),
 *     SerenityBDDReporter.fromJSON({
 *       specDirectory: './features'            // optional configuration
 *     })
 *   ],
 * })
 * ```
 *
 * ## Using Serenity BDD Reporter with Playwright Test
 *
 * ```ts
 * // playwright.config.ts
 * import { devices } from '@playwright/test';
 * import type { PlaywrightTestConfig } from '@serenity-js/playwright-test';
 *
 * const config: PlaywrightTestConfig = {
 *
 *   reporter: [
 *     [ 'line' ],
 *     [ 'html', { open: 'never' } ],
 *     [ '@serenity-js/playwright-test', {
 *       crew: [
 *         '@serenity-js/serenity-bdd',
 *         [ '@serenity-js/core:ArtifactArchiver', { outputDirectory: 'target/site/serenity' } ],
 *       ]
 *     }]
 *   ],
 * }
 * ```
 *
 * ## Using Serenity BDD Reporter with WebdriverIO
 *
 * ```ts
 * // wdio.conf.ts
 * import { WebdriverIOConfig } from '@serenity-js/webdriverio';
 *
 * export const config: WebdriverIOConfig = {
 *
 *   framework: '@serenity-js/webdriverio',
 *
 *   serenity: {
 *     crew: [
 *       '@serenity-js/serenity-bdd',
 *        [ '@serenity-js/core:ArtifactArchiver', { outputDirectory: 'target/site/serenity' } ],
 *     ],
 *     // other Serenity/JS config
 *   },
 *   // other Protractor config
 * }
 * ```
 *
 * ## Using Serenity BDD Reporter with Protractor
 *
 * ```js
 * // protractor.conf.js
 * exports.config = {
 *   framework:     'custom',
 *   frameworkPath: require.resolve('@serenity-js/protractor/adapter'),
 *
 *   serenity: {
 *     crew: [
 *         '@serenity-js/serenity-bdd',
 *          [ '@serenity-js/core:ArtifactArchiver', { outputDirectory: 'target/site/serenity' } ],
 *     ],
 *     // other Serenity/JS config
 *   },
 *
 *   // other Protractor config
 * }
 * ```
 *
 * ## Configuring Serenity BDD Reporter
 *
 * To override Serenity BDD Reporter default configuration, provide a [`SerenityBDDReporterConfig`](https://serenity-js.org/api/serenity-bdd/interface/SerenityBDDReporterConfig/)
 * as the second element of the [`SerenityBDDReporterConfig`](https://serenity-js.org/api/serenity-bdd/interface/SerenityBDDReporterConfig/) array
 * using your test runner-specific configuration mechanism.
 *
 * For example, to change the default location
 * of the [requirements hierarchy root directory](https://serenity-bdd.github.io/docs/reporting/living_documentation#the-requirements-hierarchy),
 * specify the `specDirectory` property:
 *
 * ```js
 *     crew: [
 *       [ '@serenity-js/serenity-bdd', { specDirectory: './features' } ],
 *       // ...
 *     ],
 * ```
 *
 * ### Learn more:
 * - [Serenity BDD Reporter integration documentation](https://serenity-js.org/handbook/reporting/serenity-bdd-reporter/)
 * - [Serenity/JS examples on GitHub](https://github.com/serenity-js/serenity-js/tree/main/examples)
 *
 * @group Stage
 */
class SerenityBDDReporter {
    processors;
    stage;
    eventQueues = new core_1.DomainEventQueues();
    static fromJSON(config) {
        return new SerenityBDDReporterBuilder(config);
    }
    /**
     * @param {EventQueueProcessors} processors
     * @param {Stage} [stage]
     *  The stage this [`StageCrewMember`](https://serenity-js.org/api/core/interface/StageCrewMember/) should be assigned to
     */
    constructor(processors, stage) {
        this.processors = processors;
        this.stage = stage;
        (0, tiny_types_1.ensure)('processors', processors, (0, tiny_types_1.isDefined)());
    }
    /**
     * @inheritDoc
     */
    assignedTo(stage) {
        this.stage = stage;
        return this;
    }
    /**
     * @inheritDoc
     */
    notifyOf(event) {
        if (this.isSceneSpecific(event)) {
            this.eventQueues.enqueue(event);
        }
        else if (event instanceof events_1.TestRunFinishes) {
            const id = model_1.CorrelationId.create();
            this.stage.announce(new events_1.AsyncOperationAttempted(new model_1.Name(this.constructor.name), new model_1.Description(`Generating Serenity BDD JSON reports...`), id, this.stage.currentTime()));
            try {
                this.processors
                    .process(this.eventQueues)
                    .forEach(result => {
                    this.stage.announce(new events_1.ArtifactGenerated(result.sceneId, result.name, result.artifact, this.stage.currentTime()));
                });
                this.stage.announce(new events_1.AsyncOperationCompleted(id, this.stage.currentTime()));
            }
            catch (error) {
                this.stage.announce(new events_1.AsyncOperationFailed(error, id, this.stage.currentTime()));
            }
        }
    }
    isSceneSpecific(event) {
        return Object.prototype.hasOwnProperty.call(event, 'sceneId');
    }
}
exports.SerenityBDDReporter = SerenityBDDReporter;
class SerenityBDDReporterBuilder {
    config;
    constructor(config) {
        this.config = config;
    }
    build({ stage, fileSystem }) {
        (0, tiny_types_1.ensure)('stage', stage, (0, tiny_types_1.isDefined)());
        (0, tiny_types_1.ensure)('fileSystem', fileSystem, (0, tiny_types_1.isDefined)());
        const userDefinedSpecDirectory = this.config.specDirectory && io_1.Path.from(this.config.specDirectory);
        const reporterConfig = {
            includeAbilityDetails: this.config?.reporter?.includeAbilityDetails ?? true,
            ...this.config?.reporter,
        };
        const requirementsHierarchy = new io_1.RequirementsHierarchy(fileSystem, userDefinedSpecDirectory);
        const processors = new processors_1.EventQueueProcessors(requirementsHierarchy, reporterConfig);
        return new SerenityBDDReporter(processors, stage);
    }
}
